ethers-solc/src/artifact_output.rs

```diff
@@ -0,0 +1,495 @@
+
+        Ok(artifacts)
+    }
+
+    /// Writes additional files for the contracts if the included in the `Contract`, such as `ir`,
+    /// `ewasm`, `iropt`.
+    ///
+    /// By default, these fields are _not_ enabled in the [`crate::Settings`], see
+    /// [`crate::Settings::default_output_selection()`], and the respective fields of the
+    /// [`Contract`] will `None`. If they'll be manually added to the `output_selection`, then
+    /// we're also creating individual files for this output, such as `Greeter.iropt`,
+    /// `Gretter.ewasm`
+    fn write_extras(contracts: &VersionedContracts, layout: &ProjectPathsConfig) -> Result<()> {
+        for (file, contracts) in contracts.as_ref().iter() {
+            for (name, versioned_contracts) in contracts {
+                for c in versioned_contracts {
+                    let artifact_path = if versioned_contracts.len() > 1 {
+                        Self::output_file_versioned(file, name, &c.version)
+                    } else {
+                        Self::output_file(file, name)
+                    };
+
+                    let file = layout.artifacts.join(artifact_path);
+                    utils::create_parent_dir_all(&file)?;
+
+                    if let Some(iropt) = &c.contract.ir_optimized {
+                        fs::write(&file.with_extension("iropt"), iropt)
+                            .map_err(|err| SolcError::io(err, file.with_extension("iropt")))?
+                    }
+
+                    if let Some(ir) = &c.contract.ir {
+                        fs::write(&file.with_extension("ir"), ir)
+                            .map_err(|err| SolcError::io(err, file.with_extension("ir")))?
+                    }
+
+                    if let Some(ewasm) = &c.contract.ewasm {
+                        fs::write(
+                            &file.with_extension("ewasm"),
+                            serde_json::to_vec_pretty(&ewasm)?,
+                        )
+                        .map_err(|err| SolcError::io(err, file.with_extension("ewasm")))?;
+                    }
+
+                    if let Some(evm) = &c.contract.evm {
+                        if let Some(asm) = &evm.assembly {
+                            fs::write(&file.with_extension("asm"), asm)
+                                .map_err(|err| SolcError::io(err, file.with_extension("asm")))?
+                        }
+                    }
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Returns the file name for the contract's artifact
+    /// `Greeter.json`
+    fn output_file_name(name: impl AsRef<str>) -> PathBuf {
+        format!("{}.json", name.as_ref()).into()
+    }
+
+    /// Returns the file name for the contract's artifact and the given version
+    /// `Greeter.0.8.11.json`
+    fn output_file_name_versioned(name: impl AsRef<str>, version: &Version) -> PathBuf {
+        format!("{}.{}.{}.{}.json", name.as_ref(), version.major, version.minor, version.patch)
```

Comment:
May have mentioned this elsewhere too, should we be storing the optimizer runs too?

Reply:
this info is part of the CacheEntry in the cache file, but now that we can also emit ewasm, it, asm files, perhaps we should also store this in a separate file?